design/ceph/osd-migration.md

```diff
@@ -0,0 +1,230 @@
+	OSDs                *ReplaceQueue    `json:"osds"`
+	IsPgHealthy          bool            `json:"isPgHealthy"`
+	OSDInProgress        *OSDReplaceInfo `json:"OSDInProgress"`
+	StartTimeStamp       time.Time       `json:"startTimeStamp"`
+}
+```
+
+**OSDReplaceInfo**: Represents an OSD that needs to be replaced
+  - **ID**: OSD ID of the OSD to be replaced.
+  - **Path**: Disk or pvc claim that the OSD is consuming.
+  - **Node**: Node on which the OSD is running
+
+
+**ReplaceQueue**: This is a queue implementation containing a list of OSDs to be replaced. It will support basic queue operations like push, pop, size etc.
+
+
+**OSDReplaceConfig**: This will hold the current status/progress of the OSD replacement in the cluster.
+  - **OSDs**:  List of OSDs that are yet to be replaced.
+  - **IsPgHealth**: Whether Placement Groups are healthy or not. This is a boolean field.
+  - **OSDInProgress**: Represents the OSDs for which the deployment is being deleted.
+  - **StartTimeStamp**: Time at which the whole OSD replacement process was started. This will help users know how much time has elapsed since the first OSD was replaced.
+
+
+#### Replace OSD
+
+Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label not matching the osd store type provided in `spec.storage.store.type`. This list will be stored in `OSDReplaceConfig.OSDs`
+  2. If the PG status is not `active+clean`, don't replace any OSD.
+  3. If the PG status is `active+clean`:
+     - If `OSDInProgress` is not empty due to any error in the previous reconcile, then don't remove any new OSD from the `OSDReplaceConfig.OSDs` and continue with the deletion of the OSD mentioned in `OSDInProgress`.
+     - Else remove an OSD entry from the `OSDReplaceConfig.OSDs` and assign it to `OSDInProgress`.
+  4. Delete the OSD deployment
+  5. Provide a new environment variables to the OSD prepare pod running on `OSDReplaceInfo.Node` in case of OSD on disks or `OSDReplaceInfo.Path` in case of OSD on PVCs
+      - `ROOK_REPLACE_OSD_ID`: OSD to be deleted.
+  6. OSD prepare pod job will destroy the OSD based on the environment variables provided above. This process is mentioned in the `Destroy OSD` steps below.
+  7. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDReplaceConfig.OSDs`. Skipping update operation on the OSDs that need to be migrated is important because with OSDs on disks, the OSD prepared pod is updated if there is an update in the spec. This will update the osd store to `bluestore-rdr` in the spec without migrating the OSD. This will cause errors.
+  8. Once the destroy OSD pod is added back, reset `OSDInProgress` to nil. If there is any error during the OSD migration, then don't reset `OSDInProgress`.
```

Comment:
If the OSDInProgress is the only thing in the configmap, we could just delete the configmap.